server/src/main/java/com/google/codeu/servlets/BlobStoreServlet.java

```diff
@@ -0,0 +1,119 @@
+    if (!userService.isUserLoggedIn()) {
+      response.sendRedirect("/");
+      return;
+    }
+
+    String user = userService.getCurrentUser().getEmail();
+    String userText = Jsoup.clean(request.getParameter("text"), Whitelist.none());
+
+    String regex = "(https?://\\S+\\.(png|jpg))";
+    String replacement = "<img src=\"$1\" />";
+    String textWithImagesReplaced = userText.replaceAll(regex, replacement);
+
+    Message message = new Message(user, textWithImagesReplaced);
+    datastore.storeMessage(message);
+
+    response.sendRedirect("/userpage?user=" + user);
+    String user = userService.getCurrentUser().getEmail();
+    String text = Jsoup.clean(request.getParameter("text"), Whitelist.none());
+
+    List<String> imageBlobUrls = getUploadedFileUrl(request, "image");
+    if (imageBlobUrls != null) {
+      for (String u : imageBlobUrls) {
+        text += String.format("<br/><img src=\"%s\"/>", u);
+      }
+    }
+
+    Message message = new Message(user, text);
+    datastore.storeMessage(message);
+
+    response.sendRedirect("/user-page.html?user=" + user);
+  }
+
+  /** should send a get request to display img in url. */
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    String uploadUrl = blobstoreService.createUploadUrl("/messages");
+
+    response.setContentType("text/html");
+    response.getOutputStream().println(uploadUrl);
+  }
+
+  private List<String> getUploadedFileUrl(HttpServletRequest request, String formInputElementName) {
+    BlobstoreService blobstoreService = BlobstoreServiceFactory.getBlobstoreService();
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(formInputElementName);
+
+    // User submitted form without selecting a file, so we can't get a URL.
+    // (devserver)
+    if (blobKeys == null || blobKeys.isEmpty()) {
+      System.out.printf(
+          "Assuming this is the devserver and nothing was uploaded; blobKeys: %s%n", blobKeys);
+      return null;
+    }
+
+    for (BlobKey blobKey : blobKeys) {
+      BlobInfo blobInfo = new BlobInfoFactory().loadBlobInfo(blobKey);
+      if (blobInfo.getSize() == 0) {
+        System.out.printf("deleting blobkey [%s] because it's empty", blobKey);
+        blobstoreService.delete(blobKey);
+      }
+    }
+
+    // Use ImagesService to get a URL that points to the uploaded file.
+    ImagesService imagesService = ImagesServiceFactory.getImagesService();
+    List<String> imageBlobUrls = new ArrayList<>();
+    for (BlobKey blobK : blobKeys) {
+      ServingUrlOptions options = ServingUrlOptions.Builder.withBlobKey(blobK);
+      // Checking the validity of the file to make sure it's an image, if not catch
+      // exception.
+      try {
+        // getting the image URL to the uploaded file
```

Comment:
nit: Capitalize first letter of comment.

`// getting the image URL to the uploaded file`